staging/src/k8s.io/apiserver/pkg/endpoints/filters/authn_audit.go

```diff
@@ -0,0 +1,187 @@
+}
+
+// NewAuditFailedAuthenticationDecorator returns decorator responsible for
+// logging failed authentication attempts.
+func NewAuditFailedAuthenticationDecorator(requestContextMapper request.RequestContextMapper, sink audit.Sink, policy policy.Checker) *auditFailedAuthenticationDecorator {
+	if sink == nil || policy == nil {
+		return nil
+	}
+	return &auditFailedAuthenticationDecorator{
+		requestContextMapper: requestContextMapper,
+		sink:                 sink,
+		policy:               policy,
+	}
+}
+
+var _ FailedAuthenticationDecorator = &auditFailedAuthenticationDecorator{}
+
+type auditFailedAuthenticationDecorator struct {
+	requestContextMapper request.RequestContextMapper
+	sink                 audit.Sink
+	policy               policy.Checker
+}
+
+func (f *auditFailedAuthenticationDecorator) Decorate(rw http.ResponseWriter, user user.Info, req *http.Request) http.ResponseWriter {
+	delegate := &auditFailedAuthnResponseWriter{
+		ResponseWriter:       rw,
+		user:                 user,
+		request:              req,
+		requestContextMapper: f.requestContextMapper,
+		sink:                 f.sink,
+		policy:               f.policy,
+	}
+	// check if the ResponseWriter we're wrapping is the fancy one we need
+	// or if the basic is sufficient
+	_, cn := rw.(http.CloseNotifier)
+	_, fl := rw.(http.Flusher)
+	_, hj := rw.(http.Hijacker)
+	if cn && fl && hj {
+		return &fancyFailedAuthnResponseWriterDelegator{delegate}
+	}
+	return delegate
+}
+
+var _ http.ResponseWriter = &auditFailedAuthnResponseWriter{}
+
+// auditFailedAuthnResponseWriter intercepts WriteHeader, sets it in the event.
+type auditFailedAuthnResponseWriter struct {
+	http.ResponseWriter
+	event                *auditinternal.Event
+	once                 sync.Once
+	user                 user.Info
+	request              *http.Request
+	requestContextMapper request.RequestContextMapper
+	sink                 audit.Sink
+	policy               policy.Checker
+}
+
+func (a *auditFailedAuthnResponseWriter) setHttpHeader() {
+	if a.event != nil {
+		a.ResponseWriter.Header().Set(auditinternal.HeaderAuditID, string(a.event.AuditID))
+	}
+}
+
+func (a *auditFailedAuthnResponseWriter) Write(bs []byte) (int, error) {
+	// the Go library calls WriteHeader internally if no code was written yet. But this will go unnoticed for us
+	a.processCode(http.StatusOK)
+	a.setHttpHeader()
+	return a.ResponseWriter.Write(bs)
+}
+
+func (a *auditFailedAuthnResponseWriter) WriteHeader(code int) {
+	a.processCode(code)
+	a.setHttpHeader()
+	a.ResponseWriter.WriteHeader(code)
+}
+
+func (a *auditFailedAuthnResponseWriter) processCode(code int) {
+	a.once.Do(func() {
+		ctx, ok := a.requestContextMapper.Get(a.request)
+		if !ok {
+			responsewriters.InternalError(a.ResponseWriter, a.request, errors.New("no context found for request"))
+			return
+		}
+
+		// if there already exists an audit event in the context we don't need to do anything
+		if ae := request.AuditEventFrom(ctx); ae != nil {
```

Comment:
I left this check here, but it shouldn't be needed, since we hook in failed authn specifically.